class.c

```diff
@@ -0,0 +1,1035 @@
+            if(PL_op->op_flags & OPf_STACKED) {
+                SV **svp = PL_stack_base + POPMARK + 1;
+                STRLEN svcount = SP - svp + 1;
+
+                if(svcount % 2)
+                    Perl_warner(aTHX_
+                            packWARN(WARN_MISC), "Odd number of elements in hash field initialization");
+
+                while(svp <= SP) {
+                    SV *key = *svp; svp++;
+                    SV *val = svp <= SP ? *svp : &PL_sv_undef; svp++;
+
+                    hv_store_ent(hv, key, newSVsv(val), 0);
+                }
+            }
+            val = (SV *)hv;
+            break;
+        }
+    }
+
+    fields[fieldix] = val;
+
+    PADOFFSET padix = PL_op->op_targ;
+    if(padix) {
+        SAVESPTR(PAD_SVl(padix));
+        SV *sv = PAD_SVl(padix) = SvREFCNT_inc(val);
+        save_freesv(sv);
+    }
+
+    RETURN;
+}
+
+XS(injected_constructor);
+XS(injected_constructor)
+{
+    dXSARGS;
+
+    HV *stash = (HV *)XSANY.any_ptr;
+    assert(HvSTASH_IS_CLASS(stash));
+
+    struct xpvhv_aux *aux = HvAUX(stash);
+
+    if((items - 1) % 2)
+        Perl_warn(aTHX_ "Odd number of arguments passed to %" HvNAMEf_QUOTEDPREFIX " constructor",
+                HvNAMEfARG(stash));
+
+    HV *params = NULL;
+    {
+        /* Set up params HV */
+        params = newHV();
+        SAVEFREESV((SV *)params);
+
+        for(I32 i = 1; i < items; i += 2) {
+            SV *name = ST(i);
+            SV *val  = (i+1 < items) ? ST(i+1) : &PL_sv_undef;
+
+            /* TODO: think about sanity-checking name for being 
+             *   defined
+             *   not ref (but overloaded objects?? boo)
+             *   not duplicate
+             * But then,  %params = @_;  wouldn't do that
+             */
+
+            hv_store_ent(params, name, SvREFCNT_inc(val), 0);
+        }
+    }
+
+    SV *instance = newSVobject(aux->xhv_class_next_fieldix);
+    SvOBJECT_on(instance);
+    SvSTASH_set(instance, MUTABLE_HV(SvREFCNT_inc_simple(stash)));
+
+    SV *self = sv_2mortal(newRV_noinc(instance));
+
+    assert(aux->xhv_class_initfields_cv);
+    {
+        ENTER;
+        SAVETMPS;
+
+        EXTEND(SP, 2);
+        PUSHMARK(SP);
+        PUSHs(self);
+        if(params)
+            PUSHs((SV *)params); // yes a raw HV
+        else
+            PUSHs(&PL_sv_undef);
+        PUTBACK;
+
+        call_sv((SV *)aux->xhv_class_initfields_cv, G_VOID);
+
+        SPAGAIN;
+
+        FREETMPS;
+        LEAVE;
+    }
+
+    if(aux->xhv_class_adjust_blocks) {
+        CV **cvp = (CV **)AvARRAY(aux->xhv_class_adjust_blocks);
+        U32 nblocks = av_count(aux->xhv_class_adjust_blocks);
+
+        for(U32 i = 0; i < nblocks; i++) {
+            ENTER;
+            SAVETMPS;
+            SPAGAIN;
+
+            EXTEND(SP, 2);
+
+            PUSHMARK(SP);
+            PUSHs(self);  /* I don't believe this needs to be an sv_mortalcopy() */
+            PUTBACK;
+
+            call_sv((SV *)cvp[i], G_VOID);
```

Comment:
Don't we need `SPAGAIN` *after* call_sv?

Reply:
@Ovid:
> 1. `:isa(Some::Class)` doesn't `require Some::Class`

WFM:

```
$ cat BaseClass.pm 
use v5.36;
use feature 'class';
class BaseClass { }
say "Now loaded BaseClass.pm";
1;

$ ./perl -Ilib -I. -E 'use v5.36; use feature "class"; class Derived :isa(BaseClass) { }'
class is experimental at -e line 1.
class is experimental at BaseClass.pm line 3.
Now loaded BaseClass.pm
```

> 2. Forward declarations of methods cause segfaults

Fixed.

> 3. `__CLASS__` and or `$class` need to be available in classes

`__CLASS__` to be added later.

> 4. `Data::Dumper` and related tools need to be updated to handle `PVOBJ`

Much thinking is ongoing about what's best there. There's no conclusion yet.

> 5. Unicode identifiers need to be addressed

I believe that's now fixed.

> 6. Miscellaneous segfaults and memory wrap errors when doing strange things with non-scalar fields

I'll look into this one now.

Reply:
Ahyes. Now added.